From 1be41c374af50a8dbb9ecc272ce1555d800a645b Mon Sep 17 00:00:00 2001
From: Winson Chung <winsonc@google.com>
Date: Fri, 18 Oct 2024 04:25:48 +0000
Subject: [PATCH] Remove user TIS unlocked runnable when a TIS instance is
 destroyed

- If the TIS instance is destroyed, then we should remove any queued
  user-unlocked runnables to ensure that they do not attempt to run
  again

Flag: EXEMPT bugfix
Bug: 373671447
Test: atest NexusLauncherTests
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:a9355d4dba3f0bf93cb1a28eb738c6c05dc43ca9)
Merged-In: I8ca3cdfa6f849bce5d347f14038e1ebd6bc6ff06
Change-Id: I8ca3cdfa6f849bce5d347f14038e1ebd6bc6ff06

(cherry picked from commit 4dfb35acc147280c15ed14d1d8458309e33f805e)

Backport to LineageOS 21: adjust context around the existing swipe handler and
service cleanup; keep the original callback cancellation and taskbar ordering.
---
 .../com/android/quickstep/TouchInteractionService.java | 10 ++++++++--
 src/com/android/launcher3/util/LockedUserState.kt      |  7 +++++++
 2 files changed, 15 insertions(+), 2 deletions(-)

diff --git a/quickstep/src/com/android/quickstep/TouchInteractionService.java b/quickstep/src/com/android/quickstep/TouchInteractionService.java
index b088c017db..b2b8153ebb 100644
--- a/quickstep/src/com/android/quickstep/TouchInteractionService.java
+++ b/quickstep/src/com/android/quickstep/TouchInteractionService.java
@@ -464,6 +464,10 @@ public class TouchInteractionService extends Service {
     private final AbsSwipeUpHandler.Factory mFallbackSwipeHandlerFactory =
             this::createFallbackSwipeHandler;
 
+    // This needs to be a member to be queued and potentially removed later if the service is
+    // destroyed before the user is unlocked
+    private final Runnable mUserUnlockedRunnable = this::onUserUnlocked;
+
     private final ScreenOnTracker.ScreenOnListener mScreenOnListener = this::onScreenOnChanged;
 
     private ActivityManagerWrapper mAM;
@@ -502,8 +506,7 @@ public class TouchInteractionService extends Service {
         BootAwarePreloader.start(this);
 
         // Call runOnUserUnlocked() before any other callbacks to ensure everything is initialized.
-        LockedUserState.get(this).runOnUserUnlocked(this::onUserUnlocked);
-        LockedUserState.get(this).runOnUserUnlocked(mTaskbarManager::onUserUnlocked);
+        LockedUserState.get(this).runOnUserUnlocked(mUserUnlockedRunnable);
         mDeviceState.addNavigationModeChangedCallback(this::onNavigationModeChanged);
         sConnected = true;
 
@@ -567,6 +570,8 @@ public class TouchInteractionService extends Service {
 
         mOverviewComponentObserver.setOverviewChangeListener(this::onOverviewTargetChange);
         onOverviewTargetChange(mOverviewComponentObserver.isHomeAndOverviewSame());
+
+        mTaskbarManager.onUserUnlocked();
     }
 
     public OverviewCommandHelper getOverviewCommandHelper() {
@@ -667,6 +672,7 @@ public class TouchInteractionService extends Service {
         mTaskbarManager.destroy();
         sConnected = false;
 
+        LockedUserState.get(this).removeOnUserUnlockedRunnable(mUserUnlockedRunnable);
         ScreenOnTracker.INSTANCE.get(this).removeListener(mScreenOnListener);
         super.onDestroy();
     }
diff --git a/src/com/android/launcher3/util/LockedUserState.kt b/src/com/android/launcher3/util/LockedUserState.kt
index 0a87594fde..2bef09c24a 100644
--- a/src/com/android/launcher3/util/LockedUserState.kt
+++ b/src/com/android/launcher3/util/LockedUserState.kt
@@ -72,6 +72,13 @@ class LockedUserState(private val mContext: Context) : SafeCloseable {
         mUserUnlockedActions.add(action)
     }
 
+    /**
+     * Removes a previously queued `Runnable` to be run when the user is unlocked.
+     */
+    fun removeOnUserUnlockedRunnable(action: Runnable) {
+        mUserUnlockedActions.remove(action)
+    }
+
     companion object {
         @VisibleForTesting
         @JvmField
-- 
2.53.0

